snuba/subscriptions/scheduler_processing_strategy.py

```diff
@@ -257,3 +278,167 @@ def terminate(self) -> None:
+    scheduled topic to be picked up by the subscription executor which
+    will run the actual query later.
+
+    When a tick is submitted to this strategy, we produce all of it's
+    subscriptions and place it in a queue without waiting to see if they
+    were succesfully produced or not.
+
+    On each call to poll(), we remove completed futures from the queue.
+    If all scheduled subscription messages for that tick were succesfully
+    produced and the tick indicates it should be committed, we commit offsets.
+    """
+
+    def __init__(
+        self,
+        entity_key: EntityKey,
+        schedulers: Mapping[int, Scheduler[Subscription]],
+        producer: Producer[KafkaPayload],
+        scheduled_topic_spec: KafkaTopicSpec,
+        commit: Callable[[Mapping[Partition, Position]], None],
+    ) -> None:
+        self.__schedulers = schedulers
+        self.__encoder = SubscriptionScheduledTaskEncoder(entity_key)
+        self.__producer = producer
+        self.__scheduled_topic = Topic(scheduled_topic_spec.topic_name)
+        self.__commit = commit
+        self.__closed = False
+
+        # Stores each tick with it's futures
+        self.__queue = ScheduledSubscriptionQueue()
+
+        # Not a hard max
+        self.__max_buffer_size = 10000
+
+    def poll(self) -> None:
+        # Remove completed tasks from the queue and raise if an exception occurred.
+        # This method does not attempt to recover from any exception.
+        # Also commits any offsets required.
+        while True:
+            tick_subscription = self.__queue.peek()
+
+            if tick_subscription is None:
+                break
+
+            if not tick_subscription.subscription_future.done():
+                break
+
+            tick_subscription.subscription_future.result()
+
+            self.__queue.popleft()
+
+            if tick_subscription.should_commit:
+                self.__commit(
+                    {
+                        tick_subscription.tick_message.partition: Position(
+                            tick_subscription.tick_message.offset,
+                            tick_subscription.tick_message.timestamp,
+                        )
+                    }
+                )
+
+    def submit(self, message: Message[CommittableTick]) -> None:
+        assert not self.__closed
+
+        # If queue is full, raise MessageRejected to tell the stream
+        # processor to pause consuming
+        if len(self.__queue) >= self.__max_buffer_size:
+            raise MessageRejected
+
+        # Otherwise, add the tick message and all of it's subscriptions to
+        # the queue
+        tick = message.payload.tick
+        tasks = self.__schedulers[tick.partition].find(tick.timestamps)
+        self.__queue.append(
+            message,
+            deque(
+                [
+                    self.__producer.produce(
+                        self.__scheduled_topic, self.__encoder.encode(task)
+                    )
```

Comment:
I think we should partition the subscription by subscription_id so that the same subscription id is always going to the same executor and they are executed sequentially (so the sentry consumer does not receive data points out of order).

Reply:
It is partitioned by subscription_id. It's done by the subscription task codec here https://github.com/getsentry/snuba/blob/badf1a0f2126af1da29712cf27abb5aabe8d521c/snuba/subscriptions/codecs.py#L73